### PR TITLE
Marketplace: Add "Add to Store" button for free and WordPress.org products #43616 

### PR DESCRIPTION
**Pull request by raicem**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Allows merchants to install any free and .org hosted product from any WooCommerce In-App Marketplace page, with a new **Add to Store** button on product cards.   
    - This flows looks at Woo.com search response, determines which products are installable, checks store and user capabilities to show the button. 
    - Clicking the button creates an order on Woo.com and the product is downloaded to the store.

Almost all of the code is reviewed at #https://github.com/woocommerce/woocommerce/pull/43616. This branch has extra tweaks.

### How to test the changes in this Pull Request:

**Testing the happy path**
1. Build assets with `pnpm --filter='@woocommerce/plugin-woocommerce' build`
2. Disconnect store from Woo.com by clicking **Avatar > Disconnect** from store (located at the top right of the page)
    - Please note which page you initiate the disconnection from. If you clicked **Disconned** when on the **Featured** tab, you should be taken back to that tab.
3. [On the search page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions), find a product with the **Add to Store** button and click it. It should show a warning for requiring a connection.
    - The Woo.com Search API sends an `is_installable` flag for WordPress.org products. Core takes this flag and runs it's own checks to show this button. Button will show up if:
        - We're showing plugin card and not a theme.
        - We're not on the **Featured** tab.
        - User has `install_plugins` capability.
        - Store doesn't have this plugin installed already.
5. Clicking **Connect** button should take you to Woo.com where you need to login and approve connection. After doing that, you'll be taken back to the store. Please note when you come back to your store, the install modal should open automatically.
6. In the next modal, click **Install**. This will first ping Woo.com, create an order, and then start downloading the product from the WordPress.org repository. 
    - If this is a brand new Woo.com account and it doesn't have billing address information, you'll be redirected to Woo.com checkout, with the product in your cart. 
    - The ping to Woo.com actually goes through store's backend first. So it's kinda like proxied. This is for safety and authentication. The endpoint is defined at `plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php`. 
    - If the order is complete within In-App Marketplace, you can check it on https://woo.com/my-account/orders/.
    - Downloading is done with `window.wp.updates` script, we are using this on the **My Subscriptions** and it's a core WordPress functionality (https://github.com/woocommerce/woocommerce/pull/40752/files).
    - :warning: I'm thinking downloading the zip file is not a guarantee depending on the host. That's handled but we'll show a generic error message.
8. When the installation is done and we're allowed to activate, you should see the **Activate** button. Clicking the button will activate the product and then show an option to see the plugin list, to set up the plugin. 
    - :warning: I have seen some cases where you can download but cannot activate. In these cases, we show a generic error message.
        - If the plugin is installed previously, but deleted without deactivating. 
        - You don't have the right capability for activating.
9. When you close the modal, you should no longer see **Add to Store** button on the product you have just installed.

**Testing the redirect**
11. On Woo.com, delete your customer's billing address and try to install the plugin again. 
12. Trying to install a new product, should first show a warning and then redirect user to Woo.com checkout with the product added to cart.

**Potential edge cases**
15. The `install_plugins` capability. You can create a new shop manager user and see how it works for them. I think that user is missing that capability. 
16. Maybe Woo.com is down? Search won't work at all in that case, but they might see a cache. If that's the case, they'll see a generic error on the modal. I tested thsi by blocking access to certain Woo.com URLs in browser console. 

**For Team Fire to test feel free to ignore - tested previously at #https://github.com/woocommerce/woocommerce/pull/43616**
18. When it's done, we should test installations with a product key. Find a product key from the API requests in [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) and then visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&install=PRODUCT_KEY_STARTS_WITH_W00.
19. On the [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) click **Install** in the table and confirm it works.  

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### Screenshots
**Add to Store button**
image

**Install Modal**
image

### CODE REVIEW

**Review by bgrgicak — APPROVED** [review]
Should we hide the rating?

Screenshot 2024-01-24 at 07 33 43

**Review by kdevnel — COMMENTED** [review]
When I clicked to connect my [Woo.com](http://woo.com/) account, I was redirected to [Woo.com](http://woo.com/) and then got an error after clicking the “Connect” button.

`Something went wrong. Could not obtain connected user info in _helper_auth_return.`

After a refresh, I couldn't replicate. The rest of the flow looks great!

**Comment by alopezari** [comment]
Thanks for raising importance of this PR through the label `needs: analysis`, @raicem! We will keep it in mind at release testing. I've re-run the failing tests in case they are flaky or they could potentially be failing because any of the changes from this PR, will keep you posted 👍 

I'll remove the `needs: analysis` label, it will get added automatically once the PR is merged into `trunk`.

Update: Tests passed on re-run 👍

**Comment by raicem** [comment]
Thanks @alopezari for checking it! Do you want me to merge this to trunk today? Or is that something you do as part of testing. 

@bgrgicak I [fixed](https://github.com/woocommerce/woocommerce/pull/43983/commits/97e61a6ac776b3769aafcf8430e3012ee9c444e7) the ratings issue you have mentioned. I couldn't replicate it, but reading the code, there might be a case I haven't considered.

**Comment by alopezari** [comment]
My pleasure @raicem!

Ideally, it should be tested by the PR author & reviewer (4 eyes are better than 2) prior to the merge, using the testing instructions as a general guide but not necessarily sticking to it (in fact it would be great if you could also test other edge cases), and you are totally free to merge this PR once you feel it's ready for merge.

Then, we will also make sure to test it in the release/pre-release package prior to the actual release, which is useful too for finding potential system integration issues that might occur as part of adding new multiple PRs all together.

**Pull request merged**
